packages/astro/src/core/path.ts

```diff
@@ -38,3 +38,11 @@ export function startsWithDotSlash(path: string) {
 export function isRelativePath(path: string) {
 	return startsWithDotDotSlash(path) || startsWithDotSlash(path);
 }
+
+function isString(path: unknown): path is string {
+	return typeof path === 'string' || path instanceof String;
+}
+
+export function joinPaths(...paths: (string | undefined)[]) {
+	return paths.filter(isString).map(trimSlashes).join('/');
```

Comment:
I prefer being explicit and using `filter(x => isString(x))` rather than `filter(isString)`. Ditto with `trimSlashes`.

Reply:
Typescript doesn't figure out type narrowing with inline functions 🤔 
```js
// TS recognizes that only strings pass the filter
paths.filter(isString).map(trimSlashes)

// TS ignores the filter type and errors that `trimSlashes` is being called with `string | undefined`
paths.filter(x => isString(x)).map(x => trimSlashes(x))
```

Reply:
yup, this is oddly the "TS-approved" way to do it, afaik